SCClassLibrary/External/Ableton/LinkClock.sc

```diff
@@ -11,6 +13,169 @@ LinkClock : TempoClock {
+			}, '/LinkClock/meterReply');
+			{
+				NetAddr.broadcastFlag = true;
+				// maybe some peers had to take a different port
+				(57120 .. 57127).do { |port|
+					NetAddr("255.255.255.255", port).sendMsg('/LinkClock/queryMeter', this.beats);
+				};
+				(timeout * this.tempo).wait;  // seconds, not beats
+				resp.free;
+				action.value(replies.values);
+			}.fork(this);
+		} {
+			// no peers, don't bother asking
+			// sched is needed to make sure Conditions waiting for this method still work
+			this.sched(0, { action.value(List.new) });
+		};
+	}
+
+	adjustMeterBase { |localBeats, remoteBeats, round = 1|
+		baseBarBeat = baseBarBeat + ((localBeats - remoteBeats) % beatsPerBar).round(round);
+	}
+
+	// if there are peers with syncMeter = true, align my clock to their common barline
+	// ignore peers with syncMeter = false
+	resyncMeter { |round, verbose = true|
+		var replies, cond = Condition.new, bpbs, baseBeats, denom;
+		var newBeatsPerBar, newBase;
+		fork {
+			this.queryMeter { |val|
+				replies = val;
+				cond.unhang;
+			};
+			cond.hang;
+			replies = replies.select { |reply| reply[\syncMeter] };
+			if(replies.size > 0) {
+				// verify beatsPerBar, and myBeats - beatInBar, are common across peers
+				// (if the network is glitchy, we may have replies from different trials.
+				// but the difference between query time and remote beatInBar should be
+				// consistent.)
+				// also calculate baseBarBeat common denominator
+				bpbs = Set.new;
+				baseBeats = Set.new;
+				denom = 1;
+				replies.do { |reply, i|
+					denom = denom lcm: reply[\baseBarBeat].asFraction[1];
+					bpbs.add(reply[\beatsPerBar]);
+					baseBeats.add(reply[\queriedAtBeat] - reply[\beatInBar]);
+				};
+				if(round.isNil) { round = denom.reciprocal };
+				// Now it gets tricky. We need to reduce baseBeats based on beatsPerBar.
+				// All 'syncMeter' peers should have the same beatsPerBar (size == 1).
+				// But, if something failed, maybe there are more.
+				// So we have to do a two-step check.
+				if(bpbs.size == 1) {
+					// 'choose' = easy way to get one item from an unordered collection
+					newBeatsPerBar = bpbs.choose;
+					// Sets, by definition, cannot hold multiple items of equivalent value
+					// so this automatically collapses to minimum size!
+					baseBeats = baseBeats.collect { |x| x.round(round) % newBeatsPerBar };
+					if(baseBeats.size == 1) {
+						// do not change beats! do not ever change beats here!
+						// calculate baseBarBeat such that my local beatInBar will match theirs
+						// myBeats = replies.choose[\queriedAtBeat].round(round);
+						// theirPhase = bibs.choose;  // should be only one
+						newBase = baseBeats.choose;
+						if(verbose) { "syncing meter to %, base = %\n".postf(bpbs.choose, newBase) };
+						this.prSetMeterAtBeat(bpbs.choose, newBase);  // local only
+					} {
+						// this should not happen
+						Error("Discrepancy among 'syncMeter' Link peers; cannot sync barlines").throw;
```

Comment:
perhaps the error message should be different to distinguish it in case and severity from the one below?